### PR TITLE
Sdk delivery refactor

### PR DESCRIPTION
**Pull request by eyal-avital**
this refactor purpose is that the SDK could work also with delivery3.0.

- changed ekoPlayer constructor - construct this.ekoEmbed with ekoEmbed factory ,to expose different behaviour.
- ekoEmbed/v1 - old version api (baloo)
- ekoEmbed/v2 - new version api (delivery)

- refactor cover with cover factory

- iframeCreator.js control the iframe instanceCount

- update tests -  integration tests use puppeteer for testing ekoPlayer api

*** i still need to change package.json version***

### CODE REVIEW

File: src/EkoPlayer.js [hunk 1]
```diff
@@ -1,7 +1,11 @@
-import deepmerge from 'deepmerge';
-import EventEmitter from 'eventemitter3';
+/* eslint-disable new-cap */
+import ekoEmbedFactory from './lib/ekoEmbed/factory';
+import coverFactory from './lib/cover/coverFactory';
```
Comment (Melatonin64) [inline src/EkoPlayer.js:3]: Might be a bit nit picky, so feel free to ignore, but we should consider changing the filename for consistency (i.e. `ekoEmbed/factory` vs. `cover/coverFactory`).

File: src/EkoPlayer.js [hunk 2]
```diff
@@ -57,41 +61,28 @@ const COVER_STATES = {
...
-    STARTED: 'eko-player-started',
-};
-
-let instanceCount = 0;
 let isEkoSupported = null;
-
 class EkoPlayer {
     /**
      * Creates an instance of EkoPlayer.
      * @param {Element|string} el - The container element to be used by the player, or a DOM selector string for the container element.
```
Comment (Melatonin64) [inline src/EkoPlayer.js:68]: Missing the new param `embedapi` documentation in the JSDoc comment.

File: src/lib/ekoEmbed/factory.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import v1 from './v1';
+import v2 from './v2';
+
+const imples = {
+    '1.0': v1,
+    '2.0': v2
+};
+
+function create(iframe, embedapiVersion) {
+    const ekoEmbed = imples[embedapiVersion];
```
Comment (Melatonin64) [inline src/lib/ekoEmbed/factory.js:11]: Consider changing to `const EkoEmbedClass`

File: src/lib/ekoEmbed/v2.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        const msg = event.data;
+
+        // Do nothing if this message was not intended for us
+        if (!msg.event || msg.embedid !== this.iframe.id || msg.embedapi !== '2.0') {
+            return;
+        }
+
+        this.eventEmitter.emit(msg.event, msg.args, msg.embedid, msg.embedapi);
```
[the remark this answers is not in the document]
Comment (Melatonin64) [reply src/lib/ekoEmbed/v2.js:81]: We should also add tests to verify that both V1 and V2 behave consistently and as expected/documented...

**Review by Melatonin64 — CHANGES_REQUESTED** [review]
Overall looks really great, thanks @eyal-avital!
Let's fix these comments and do another round of review - thanks!

File: src/lib/ekoEmbed/v1.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        const msg = event.data;
+
+        // Do nothing if this message was not intended for us
+        if (!msg.type || msg.embedId !== this.iframe.id) {
+            return;
+        }
+
+        this.trigger(msg.type.replace(/^eko./, ''), msg.args);
```
Comment (Melatonin64) [inline src/lib/ekoEmbed/v1.js:85]:
```suggestion
        this.trigger(msg.type.replace(/^eko./, ''), ...msg.args);
```

Comment (Melatonin64) [reply src/lib/ekoEmbed/v1.js:85]: I think we need the spread operator here to get the desired result.
@eyal-avital can we add tests to make sure that the arguments are being passed correctly?

**Review by Melatonin64 — CHANGES_REQUESTED** [review]
Looks good to me, next steps would be:

- Suggested fix (spread operator)
- README changes - we should we document the new 2nd arg (embedapi) of the load() method in the README
- Test cover the event functionality (to make sure args are passing as expected for v1 and v2 consistently).

**Comment by DivyaMahadevan** [comment]
Looks really good! Agree with Tomer that it would be good to have some tests around args being passed correctly, but other than that it looks good to me.

File: README.md [hunk 6]
```diff
@@ -30,6 +30,7 @@ Creates an instance of an EkoPlayer.
 | Param           | Type           | Description  |
 | :-------------: |:--------------:| :------------|
 | el | `Element, String` | The container element to be used by the player, or a DOM selector string for the container element. |
+| embedapi | `1.0, 2.0` | embed api version to be used internally. |
```
Comment (Melatonin64) [inline README.md:33]:
```suggestion
| embedapi | `String` | Optional. eko embed api version to be used internally. Valid values include `"1.0"`, `"2.0"`. If no value given, default value `"1.0"` will be used. |
```

**Review by Melatonin64 — COMMENTED** [review]
Requested a tiny change in the README, otherwise looks great!!

**Review by Melatonin64 — APPROVED** [review]
Looks good to me, let's merge!
(Note to self - let's add and maintain a CHANGELOG file).

**Pull request merged**